### ISSUE TITLE
Need help fetching past and new events from a contract while also checking sync status

### ISSUE THREAD

**Issue by WyseNynja** [tomusdrw/rust-web3#204]
I am trying to subscribe to events on the [uniswap](https://uniswap.info) contracts, but am having some trouble getting any logs.

If I were using pyweb3, I would want something like https://www.reddit.com/r/UniSwap/comments/atddo2/effective_way_to_get_all_uniswap_exchange/ehp8w1z/

```
from web3.auto import w3

abi = [{"name": "NewExchange", "inputs": [{"type": "address", "name": "token", "indexed": True}, {"type": "address", "name": "exchange", "indexed": True}], "anonymous": False, "type": "event"}]

uniswap = w3.eth.contract('0xc0a47dFe034B400B47bDaD5FecDa2621de6c4d95', abi=abi)
events = uniswap.events.NewExchange.createFilter(fromBlock=6627917).get_all_entries()
token_exchange = {e.args.token: e.args.exchange for e in events}

for token, exchange in token_exchange.items():
    print(token, exchange)
```

Except instead of printing, I want to subscribe to the events of that Exchange.

I am able to get the [pubsub](https://github.com/tomusdrw/rust-web3/blob/master/examples/pubsub.rs) example working, so I know my web3 connection is correct.

I will push the code I have up to GitHub soon, but wanted to get this issue made now in case there is something simple I'm missing. Right now, my rust code prints the current block number and then prints my filter, but it never prints anything else.

```
7376907
factory_filter: Filter {
    from_block: Some(
        Number(
            6627917
        )
    ),
    to_block: None,
    address: Some(
        ValueOrArray(
            [
                0xc0a47dfe034b400b47bdad5fecda2621de6c4d95
            ]
        )
    ),
    topics: Some(
        [
            Some(
                ValueOrArray(
                    [
                        0x9d42cb017eb05bd8944ab536a8b35bc68085931dd5f4356489801453923953f9
                    ]
                )
            )
        ]
    ),
    limit: None
}
```

I've tried with and without setting the topic to keccack256("NewExchange(address,address)")

Should I be using create_logs_filter or subscribe_logs or something else? Any suggestions, or do I need a more complete example?

**Comment by WyseNynja** [comment]
Here is some example code that I think is close to working: https://github.com/WyseNynja/rust-web3-uniswap

I'd really appreciate some help getting it actually working. Thanks in advance!

**Comment by WyseNynja** [comment]
I added some example python code next to the rust code. I did this to make sure my ethereum node wasn't the problem. The python code is working as expected, but the rust code never sees any contract events.

**Comment by tomusdrw** [comment]
@WyseNynja can you run the node with `-lrpc=trace` and compare what payloads are sent by rust and python implementation?

**Comment by WyseNynja** [comment]
I'll try that. Is there a similar flag I can set on the rust side to see what payloads are being sent? Something like `RUST_LOG=web3...`?

**Comment by WyseNynja** [comment]
While waiting for a local node to finish syncing so I can watch logs more easily, I made a [branch](https://github.com/WyseNynja/uniswap-demo/blob/tokenCount/rust/src/main.rs) that uses contract calls instead of events to get the data. It is working better, but I am stumped on how to properly return in different futures in an if/else in a future: https://github.com/WyseNynja/uniswap-demo/blob/d93a639aa8d9902e59e0305ebfd8bdefbb23df62/rust/src/main.rs#L162

That code is also getting hard to follow because there is so much indenting. Am I doing something wrong here? I think the closures are actually making it harder to read. Is there a cleaner way to write my code?

**Comment by WyseNynja** [comment]
I'm not sure how to get rpc logs out of geth. I'm getting `flag provided but not defined: -lrpc`

**Comment by WyseNynja** [comment]
I tried in [another branch](https://github.com/WyseNynja/uniswap-demo/blob/tokenCount_with_functions/rust/src/main.rs) to flatten the code a bit by changing the closures into functions, but I got stuck when trying to create a bunch of futures for 1 through `tokenCount()`. I think this is a cleaner way to write this, but I couldn't get it working so I'm definitely doing it wrong.

Part of the reason I'm confused is I keep finding docs/posts about tokio instead of tokio core and futures 0.2 and 0.3 instead of futures 0.1.

**Comment by tomusdrw** [comment]
@WyseNynja I don't know about Geth, might be worth asking in the go-ethereum channel. (I suspect it's going to be something like `--vmodule='eth/*=5'`).

You can enable logs in `web3`, you need to instantiate `env_logger` in your `fn main()`, like so:
```
fn main() {
  env_logger::init();
  ...
}
```

Then you can run your example with `RUST_LOG=web3=trace` and it should log the requests as well.

**Comment by WyseNynja** [comment]
I've switched my code around to not fetching past logs. It only subscribes to new logs now and that works fine. I'm going to close since I'm not looking into this anymore.
